### ISSUE TITLE
Training and Evaluation of 

### ISSUE THREAD

**Issue by harsh-agar** [harsh-agar/ShuffleSeg#1]
Training and Evaluation of 
Deep Learning Models

By
Harsh Agarwal
	(Roll No. 2015102)	

Supervisors:
Internal
Dr. Atul Gupta
Associate Professor
PDPM IIITDM Jabalpur	External
Mrs. Sheena KP
Senior Manager
Canon ISDC, Bengaluru

 

Electronics & Communication Engineering, B. Tech 2015

INDIAN INSTITUTE OF INFORMATION TECHNOLOGY, DESIGN AND MANUFACTURING, JABALPUR

22nd September, 2018 - 09th October, 2018 
1.	Introduction

The tasks allotted to me involve studying, training and testing of the different Deep learning models. 
I worked on the training and inference of various models.
Worked on the detection of pose keypoints in an image using Associative Embedding. Also, classified the pose of a person as complex or normal based on the keypoint information generated using Associative Embedding.
Then I had to work on a human head detection problem, wherein we had to detect heads in an image using a classifier trained majorly using the generated images.
The deep learning models which I worked on included: YOLOv3 and Associative Embedding model. While all YOLO models are implemented in Darknet; Associative Embedding model has been implemented using PyTorch Library. 
2.	Progress

2.1 Keypoint Detection
Associative Embedding is a model that can be used for both multi-person pose estimation and instance segmentation and it has reported state-of-the-art performance for multi-person pose estimation.
This model can efficiently estimate the 17 body keypoints of a person which include the various body joints, as well as the facial features such as the eyes, ears and nose.
 
Fig1: Both multi-person pose estimation and instance segmentation are examples of computer vision tasks that require detection of visual elements and grouping of these elements

In the Associative Embedding paper by Alejandro Newell et al. the authors have jointly perform detection and grouping using a single-stage deep net-work trained end-to-end. They have proposed associative embedding, a novel method to represent the output of joint detection and grouping. The basic idea is to introduce, for each detection, a real number that serves as a “tag” to identify the group the detection belongs to. In other words, the tags associate each detection with other detections in the same group.
Here, the associative embedding are combined with the stacked hourglass architecture, a model for dense pixelwise prediction that consists of a sequence of modules each shaped like an hourglass (as shown in Fig. 2). Each “hourglass” has a standard set of convolutional and pooling layers that process features down to a low resolution capturing the full context of the image. Then, these features are upsampled and gradually combined with outputs from higher and higher resolutions until reaching the final output resolution. Stacking multiple hourglasses enables repeated bottom-up and topdown inference to produce a more accurate final prediction.

  
Fig2: Stacked Hourglass Architecture
Training: Each model was trained over a training set comprising of around 9000 images comprising of three object classes: Person, Bike and Helmet. The hardware used was GPU NVidia 1080. And training was conducted for around 300 epochs.
2.2 Image Clustering based on Pose of a Person
The task was to classify the pose of a person based on the keypoint information of various joints and features of the body.
Here we have formed 4 different clusters, namely:  Inverted, Bent Body, Other Complex and Normal. Since the size of humans in all the images vary quite much, hence instead of considering the regular or normalized distance between key-points for pose classification we use different angles between key-points (or joints). 
The images contained in different clusters include:
Inverted: Images where the human body is inverted.
Bent Body: Comprises of the Images when the knees are bent or the complete body is bent.
Other Complex: This comprises of all the other Images which are complex in nature but not covered in the above two clusters.
Normal: All the images which do not belong to either of the clusters mentioned above are considered in normal.
2. 3 Training a head classifier using generated images
Here, we trained a head detector model using mainly generated images and used the model to detect head in the real images. YOLO v3 model with 105 layers was used for the training a detector.
The accuracy numbers generated for the model included the values for True Positive, False Positive, True Negative, False Negative, Precision, Recall and also the Average Precision for the class Real Head along with the mean Average Precision (mAP) score of the model.
Also, the results were dumped in one of the following bins on the basis of objects identified. The bins are: Detected, Not-Detected, Weak-Localization and Misclassified.
3.	Results

3.1 Inference drawn from the training Associative Embedding models:
We used Mean Squared Error for error estimation of the detected keypoints from the ground truth of the keypoints marked manually. Thus, we determine Squared Error for each keypoint in an image and then determine the mean of all those for the mean error in an image.
The results obtained were:
Using pretrained model we got an error of 0.27
After training the model for 300 epochs the error reduced to 0.06 
3.2 Image Clustering based on Pose of a Person
The clustering was performed on 2000 images of the Leeds University dataset. The pie chart shown below displays the distribution obtained for different clustering cases.
 
3.2 Training a head classifier using generated images:
We trained the head detector YOLOv3 model for around 40000 iterations using 4900 real and 3000 generated images. The testing was done on 500 complex cases of Leeds dataset and we claimed a mAP score of 0.604 
4.	Conclusion

By the work done till now I got to know about different network architecture and there training methods. I also learned about various accuracy metrics, and used them to estimate how well the model performs
Now I am working on the visualization of the features learned by different layers in the neural network.
